src/ngx_http_lua_output.c

```diff
@@ -802,4 +802,105 @@ ngx_http_lua_flush_cleanup(void *data)
     ctx->flushing_coros--;
 }
 
+
+#ifndef NGX_LUA_NO_FFI_API
+int
+ngx_http_lua_ffi_write(ngx_http_request_t *r, const char *buf, size_t offset,
+                       size_t len, char **err)
+{
+    ngx_http_lua_ctx_t          *ctx;
+    ngx_uint_t                   size;
+    ngx_buf_t                   *b;
+    ngx_chain_t                 *cl;
+    ngx_int_t                    rc;
+    const char                  *str_sub;
+
+    if (r == NULL) {
+        *err = "no request object found";
+        return NGX_ERROR;
+    }
+
+    if (buf == NULL) {
+        *err = "no buffer found";
+        return NGX_ERROR;
+    }
+
+    size = ngx_strlen(buf);
```

Comment:
@hkmo We should not assume the first argument to `ffi_write` is a Lua string. It could be just a raw C buffer. This is a much more flexible arrangement.